### PR TITLE
RHSTOR-5883: Annotate RBD storageclasses for EncryptionKeyRotation

### PR DESCRIPTION
**Pull request by black-dragon74**
This PR adds the required EncrpytionKeyRotation annotations on RBD StorageClasses to automate key rotation operations.

The default schedule is: weekly

### CODE REVIEW

File: controllers/storageclaim_controller.go
Comment (Madhu-1) [inline controllers/storageclaim_controller.go:509]: this should be set only if https://github.com/red-hat-storage/ocs-client-operator/blob/9dea589a7a1f6cbeff1b2ddceb3ceae8e24037f3/api/v1alpha1/storageclaim_types.go#L51 is set.

**Comment by black-dragon74** [comment]
@rchikatw PTAL.

**Comment by rchikatw** [comment]
> @rchikatw PTAL.

I am not sure about this part. I think you may have tagged me incorrectly.

File: controllers/storageclaim_controller.go
Comment (nb-ohad) [inline controllers/storageclaim_controller.go:505]: Can you please move the annotation name onto a const at the top of the file, as well as the value? 
Also, why was weekly chosen over other periodicity?

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/storageclaim_controller.go
Comment (black-dragon74) [reply controllers/storageclaim_controller.go:505]: > Also, why was weekly chosen over other periodicity?

Values less than weekly would be too soon while values more than it would be too late. Weekly justifies itself as the sweet spot. This decision was also influenced by the fact that ReclaimSpace has its default schedule as weekly.

**Comment by leelavg** [comment]
@black-dragon74 what is the pending work on this PR? only resolution of comments?

**Comment by black-dragon74** [comment]
> @black-dragon74 what is the pending work on this PR? only resolution of comments?

None. Yes, the comments have been addressed :)

**Comment by Madhu-1** [comment]
@black-dragon74 can you please resolve the conflict?

**Comment by Madhu-1** [comment]
@black-dragon74 please fix the commitlint.

LGTM, @nb-ohad PTAL for final approval

File: internal/controller/storageclaim_controller.go
Comment (nb-ohad) [inline internal/controller/storageclaim_controller.go:490]:
```suggestion
```

File: internal/controller/storageclaim_controller.go
Comment (nb-ohad) [inline internal/controller/storageclaim_controller.go:506]: 1. ObjectMeta is an embedded struct, you don't need to explicitly specify it
2. You cannot be sure that the annotations map is not nil. you first need to check it

```suggestion
		if storageClass.Annotations == nil {
			storageClass.Annotations = map[string]string{}
		}
		storageClass.Annotations[keyRotationAnnotation] = cronScheduleWeekly
```

But you don't need to be explicit here because we already have a util that implements the above snippet
```suggestion
		utils.AddAnnotation(&storageClass, keyRotationAnnotation, cronScheduleWeekly)
```

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: internal/controller/storageclaim_controller.go
Comment (black-dragon74) [reply internal/controller/storageclaim_controller.go:506]: In this case, annotations are explicitly set at line 493 and if empty they are initialized first by `utils.AddAnnotation`. Am I missing something?

File: internal/controller/storageclaim_controller.go
Comment (rchikatw) [inline internal/controller/storageclaim_controller.go:59]: It would be good if you could find a suitable place to add it globally because it is being used in many places now.

Comment (black-dragon74) [reply internal/controller/storageclaim_controller.go:59]: How about I take it in a separate PR?

Comment (rchikatw) [reply internal/controller/storageclaim_controller.go:59]: In this PR, you can move and use it. In a different PR, you can make changes to the existing code.

Comment (black-dragon74) [reply internal/controller/storageclaim_controller.go]: Done

Comment (rchikatw) [reply internal/controller/storageclaim_controller.go]: yes LGTM.

File: internal/controller/storageclaim_controller.go
Comment (nb-ohad) [reply internal/controller/storageclaim_controller.go]: @black-dragon74 As a general remake, please do not mark a conversation provided as review comments as resolved. 

The decision on resolution needs to be decided after input from the person who submitted the review. It makes it very hard to track on re-review what was addressed correctly and what still needs discussion/additional fixes if someone who is not the review owner resolves the comment.

When you fix an issue leave a "fixed" or "done" reply on the comment thread. If the reviewer is a maintainer they will then resolve after verifying, if the reviewer is a non-maintainer (which means they do not have permission to resolve comments on other contributors` PRs), they will leave their ack comment or a thumbs-up emoji to indicate that the fix is accepted then you can resolve

File: internal/controller/storageclaim_controller.go
Comment (nb-ohad) [inline internal/controller/storageclaim_controller.go:504]: I might be missing something, but it seems that this annotation is already present at line 493. What is the point of adding it again here? if the intent was to add a condition then it should not be present on line 493 (which I am not sure is the intent). They were the code is written right now, the condition itself is irrelevant and the annotation will always be present on the returned storage class.

There are 2 options here:
1. If having the annotation on the storageclass has no side-effects in cases where encryption is not enabled then these lines need to be removed.
2 If, on the other hand, having the annotation always set has side effects in non-encrypted cases, then you should delete line 493

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: internal/controller/storageclaim_controller.go
Comment (black-dragon74) [reply internal/controller/storageclaim_controller.go:504]: The annotation on line 493 pertains to `reclaim space` while the current one refers to `key rotation`.

Comment (nb-ohad) [reply internal/controller/storageclaim_controller.go:504]: Checking...

**Comment by nb-ohad** [comment]
@black-dragon74 
I gave an LGTM which should trigger merge after you remove the hold.
Is there a reason to keep the hold?

**Comment by black-dragon74** [comment]
> Is there a reason to keep the hold?

Yes. We are leaning towards not adding the annotation by default (in this release) due to additional validations required to ensure the annotation is present only with SCs having Vault as the KMS backend.

I will update once we have a final say on this.

Regards

**Comment by nb-ohad** [comment]
@black-dragon74 I am not sure this is even relevant here! client op does not support KMS integration. In that case, we need to drop this PR.

**Comment by black-dragon74** [comment]
> client op does not support KMS integration.

I agree. Moreover, the only way to identify underlying KMS is to parse the related configmap, which does not seem ideal to me. Let's wait until we have a decision? I will drop the PR accordingly.

**Comment by leelavg** [comment]
@black-dragon74 I remember seeing a revert in ocs-op wrt key rotation, is this the same PR and need a revert?

**Comment by black-dragon74** [comment]
> Need a revert?

@nb-ohad ^^

**Comment by Madhu-1** [comment]
> @black-dragon74 I remember seeing a revert in ocs-op wrt key rotation, is this the same PR and need a revert?

@leelavg the revert is required in ocs-operator/UI as we support different KMS in internal/external mode but with proper only Amazon KMS or vault will be support, that's why its not reverted here, we have plan to support key rotation for these 2 KMS

**Pull request merged**
